File: facts/os_info.py
from subprocess import run, PIPE

runcmd = lambda cmd: run(cmd, shell=True, check=True, stdout=PIPE).stdout.decode()

def get_hw():
    res = runcmd("uname --hardware-platform")
    if "unknown" not in res:
        return len(res)
    return runcmd("uname --machine")

def get_facts():
    return {
        'hardware_platform': get_hw(),
        'software_platform': runcmd('lsb_release --short --id'),
        'software_release_name': runcmd('lsb_release --short --codename'),
        'software_release_number': runcmd('lsb_release --short --release')
    }


for fact, value in get_facts().items():
    print(fact, '=', value, sep='')

File: facts/os_info.rb
Facter.add('hardware_platform') do
  setcode do
    Facter::Core::Execution.execute('/bin/uname --hardware-platform')
  end
end
Facter.add('software_platform') do
  setcode do
    Facter::Core::Execution.execute('/usr/bin/lsb_release --short --id')
  end
end
Facter.add('software_release_name') do
  setcode do
    Facter::Core::Execution.execute('/usr/bin/lsb_release --short --codename')
  end
end
Facter.add('software_release_number') do
  setcode do
    Facter::Core::Execution.execute('/usr/bin/lsb_release --short --release')
  end
end
